Autosave: fix conflict between remote and local autosaves

mcsf opened the pull request:
Fixes #17500

## Description
_For now, this only includes failing E2E tests_

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

mcsf committed:
Add failing E2E tests

gziolo commented:
I can see it failing as expected but only on one of the nodes, it should fail for both suites of e2e tests (https://github.com/WordPress/gutenberg/pull/17501/checks?check_run_id=230152153):

![Screen Shot 2019-09-23 at 09 30 57](https://user-images.githubusercontent.com/699132/65408226-f7bab080-dde4-11e9-92a0-40d2347c1898.png)

mcsf committed:
Fix Promise.all idiom, make it actually parallel

Also bump sleep time by 1s to allow time for local autosave.

mcsf committed:
Don't parallelise, improve reliability

mcsf commented:
> I can see it failing as expected but only on one of the nodes, it should fail for both suites of e2e tests (https://github.com/WordPress/gutenberg/pull/17501/checks?check_run_id=230152153):

Thanks for spotting. As of 208dbbb, this should be more reliable and only failing as intended.

mcsf committed:
WIP: add e2e for clearing local autosave

gziolo commented:
I don't see any changes in the logic. Should those e2e tests pass? :)

mcsf commented:
> I don't see any changes in the logic. Should those e2e tests pass? :)

No :(

mcsf committed:
Testing: Utils: Add toggleOfflineMode

toggleOfflineMode is a wrapper for Puppeteer's page.setOfflineMode which
keeps track of the tester's intention to bring the network down. This
allows the E2E framework to ignore logged network errors when these
errors are expected.

mcsf committed:
Editor store: Controls: Add local autosave handlers

mcsf committed:
LocalAutosaveMonitor: Purge local only on successful remote autosave

* Switch to new sessionStorage handlers for local autosaves
* Purge local autosave if remote autosave is successful
* Add e2e test simulating unavailable network

mcsf committed:
WIP: Avoid server/local autosave conflicts

* Defer to server autosave if any available
* WIP commit. Clearly.

youknowriad commented on packages/e2e-tests/specs/autosave.test.js:
```diff
@@ -89,18 +95,77 @@ describe( 'autosave', () => {
+
+		// Trigger local autosave
+		await page.evaluate( () => window.wp.data.dispatch( 'core/editor' ).localAutosave() );
+		expect( await page.evaluate( () => window.sessionStorage.length ) ).toBe( 1 );
```
not sure how save is that check (sessionStorage), I mean other things could be using sessionStorage. did you consider using a selector or something?

mcsf replied:
> did you consider using a selector or something?

I did, but it felt wrong to write an impure selector that wouldn't read from `state`. Maybe that's just confirmation that the use of custom store controls is wrong. :)

youknowriad replied:
Makes sense, it's not important because it's a test anyway.

youknowriad commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -37,47 +41,17 @@ const hasSessionStorageSupport = once( () => {
  *
  * @return {Function} Callback function.
  */
 function useAutosaveCallback() {
```
This refactoring is great, it also makes me think this particular hook is not necessarily needed anymore.

mcsf replied:
Perhaps, though we would still need to use `requestIdleCallback` somewhere. And wouldn't it be strange to have it in the action creator?

mcsf replied:
Or do you just mean inline the hook in the `LocalAutosaveMonitor` component?

youknowriad replied:
Yeah, I meant inline,  but I don't feel strongly tbh

youknowriad commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -88,45 +62,74 @@ function useAutosaveNotice() {
+		remoteAutosave,
+		hasFetchedAutosave,
+	} = useSelect( ( select ) => {
+		const _postId = select( 'core/editor' ).getCurrentPostId();
```
Nit: I'd just drop the `_`  as we don't really use that pattern elsewhere.

youknowriad replied:
I guess it's because  it's already use up, I  wonder if we can rename it something else? otherwise, it's fine.

mcsf replied:
Yeah, it's to avoid collision. I think I've seen the `_` usage somewhere else recently. Happy to change to whatever is best.

youknowriad commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -88,45 +62,74 @@ function useAutosaveNotice() {
-		if ( ! hasDifference ) {
-			// If there is no difference, it can be safely ejected from storage.
-			window.sessionStorage.removeItem( postKey( postId ) );
+		{
```
Is this needed? a way to create a scope?

mcsf replied:
Yes, it was for scope, and to share the WIP quickly

youknowriad commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -140,7 +143,7 @@ function useAutosaveNotice() {
 			],
 		} );
-	}, [ postId ] );
+	}, [ postId, remoteAutosave ] );
```
This means, the function is reexecuted each time `removeAutosave` changes which can happen on each autosave I guess and  not only initial rendering, is this expected?

mcsf replied:
No, indeed, not expected. I think I left this here in my frustration with the E2E tests, when for some reason `[ postId, hasFetchedAutosave ]` didn't work. I can circle back.

youknowriad commented on packages/editor/src/store/controls.js:
```diff
@@ -23,6 +23,38 @@ export function getRegistry() {
+
+export function localAutosaveClear( postId ) {
+	window.sessionStorage.removeItem( postKey( postId ) );
+}
```
What's the reasoning for keeping these utils here and using a control. I do wonder if we should just use these functions directly from like a `autosave-utils` or something like that?

mcsf replied:
I went for controls to see how it would feel to be able to dispatch saving actions just like with regular autosave.

> `autosave-utils`

These would have to be available in E2E tests, so presumably they would be globally available. Is it worth adding a new module just for this?

I guess controls are flexible for both good and bad. What's your preference here?

youknowriad replied:
I think since it has no relation with state, I'd personally prefer just a utility file somewhere

youknowriad commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -150,20 +153,30 @@ function useAutosavePurge() {
-			window.sessionStorage.removeItem( postKey( postId ) );
+		if (
+			( lastIsDirty.current && ! isDirty ) ||
+			( lastIsAutosaving.current && ! isAutosaving && ! didError )
```
Should this be the opposite (when the autosaves finishes instead of when the autosaves starts)?

Also, what's the reasoning for clearing when the post becomes dirty?

mcsf replied:
> (when the autosaves finishes instead of when the autosaves starts)

Isn't that the case already? I guess `useRef`'s notation with `.current` is confusing to read.

> Also, what's the reasoning for clearing when the post becomes dirty?

Likewise, it clears when the post is flushed.

mcsf committed:
Defer to remote autosaves. Address feedback.

* hasFetchedAutosave should wait for user resolution

* useAutosavePurge: only rely autosaving conclusion and lack of errors,
  don't look at isDirty anymore

* Restore autosave() call in e2e test

* Inline and remove useAutosaveCallback

mcsf committed:
Rename to __experimentalLocalAutosave
